### DISCUSSION TITLE
What's the purpose of the SMT backend?

### DISCUSSION THREAD

**Discussion by div72** [py2many/py2many#550] [Q&A] [answered]
What's the point of finding out that assertions in a Python program are satisfiable? Wouldn't ensuring that opposites of assertions are unsatisfiable be much more meaningful?

**Comment by jayvdb** [comment]
> Wouldn't ensuring that opposites of assertions are unsatisfiable be much more meaningful?

I am not 100% sure exactly what you are asking there (not enough coffee to parse it..), but I suspect the SMT "language" can do what you're wanting.

> What's the purpose of the SMT backend?

Note it is actually a z3 "smt2" backend.

I am still unsure of the benefit of having it here, mostly because our tests of py->z3 transpiling start with invalid Python.

e.g.
```python
py2many> PYTHONPATH=. python3 tests/cases/equations.py 
Traceback (most recent call last):
  File "/home/jayvdb/python/py2many/tests/cases/equations.py", line 6, in <module>
    x > 2
NameError: name 'x' is not defined
```

```py
> PYTHONPATH=. python3 tests/cases/demorgan.py 
Traceback (most recent call last):
  File "/home/jayvdb/python/py2many/tests/cases/demorgan.py", line 3, in <module>
    from py2many.smt import check_sat, get_value
ImportError: cannot import name 'get_value' from 'py2many.smt' (/home/jayvdb/python/py2many/py2many/smt.py)
```

So, I went down an interesting rabbit hole of working out if we can get valid Python to do the same thing, ending up with https://github.com/jayvdb/smtfe and https://github.com/py2many/py2many/pull/513 , which works but does not yet support functions.

But even if I got that working, we've still effectively created a new language which is a bastard of Python in order to transpile into z3 .  I suspect there will never be one piece of sample (useful) Python code which will be able to transpile to z3 and another backend.

The best argument I have for keeping it here is that it demonstrates py2many being useful for a very different purpose - and this isnt a bad thing however the CI time allocated to it needs to be minimal, which https://github.com/py2many/py2many/issues/540 should help achieve.

**Reply by div72** [reply]
> I am not 100% sure exactly what you are asking there (not enough coffee to parse it..), but I suspect the SMT "language" can do what you're wanting.

Currently the SMT backend is doing check-sat, which to my knowledge searches for any model which satisfies the constraints we are asking it to solve for. If we reverse the assertions and SMT solver is able to prove that no such model exist, that means our assertions are true no matter the inputs. For example I got inspired by this backend and started [a static analyzer for V](https://github.com/div72/valve) which uses this method to prove assertions/ensure array accesses are in bound. Which is a lot more meaningful in my opinion than what py2many currently does.

> however the CI time allocated to it needs to be minimal

Z3 does not seem to take much time to run AFAICT.

**Reply by jayvdb** [reply]
The goal of smtfe was to provide more of the SMT functionality in a simple to use "frontend-end" (fe), and ideally even have exit-code support for these "scripts".  e.g. "check-sat" is implemented at https://github.com/jayvdb/smtfe/blob/main/src/smtfe/__init__.py#L18 .  

see https://github.com/jayvdb/smtfe/blob/main/tests/test_variable.py#L56 to get a quick feel for it.  Or https://github.com/py2many/py2many/pull/513/files#diff-46a43619db84bd840d1040a478de3c84ad58df3d83a624e78176c384009c1f52R12 for the changes needed to py2many tests to use smtfe in order to have valid Python tests.

But only basic arithmetic is supported in smtfe.  The needs of https://github.com/div72/valve/blob/devel/valve.v would take a while to re-implement in smtfe , and I cant see anyone wanting to write real programs in this smtfe language, as they would prefer to be using some wrapper around z3, so they can use a full programming language of their choice, like you are doing with v.

**Reply by div72** [reply]
Is there any specific reason you're making your own "front-end" instead of using something like Z3Py's generate smt?

**Reply by jayvdb** [reply]
I was trying to closely match the syntax already in the py2many test cases.

**Reply by adsharma** [reply]
@div72 that topic is addressed here: https://adsharma.github.io/pysmt/

My belief is that this syntax is much more approachable vs the z3py syntax. Semantically they're the same.

**Reply by jayvdb** [reply]
> Semantically they're the same.

You are inventing a new language that only looks like Python, and for some reason it is embedded here in a project that is a transpiler.

I believer I've done all the necessary work for it to use py2many, without it being in this repo.

So, I think it should be moved to a new repo.

**Reply by adsharma** [reply]
If you see the circbuf example here: https://adsharma.github.io/pysmt/#real-world-usage

it can't be done by separating this into a new language. It's a fundamental part of a toolchain where we are transpiling verified programs (at some point in the future when all pieces are in place).

There are folks who want to do this in the rust domain, but I don't believe it's possible to do it efficiently because of the large footprint of rust. I could be proven wrong in 10 years may be :)

**Reply by adsharma** [reply]
https://graydon2.dreamwidth.org/312681.html links to a bunch of them.
https://rust-formal-methods.github.io/

The overall system setup:

```
python source (both runtime and verification code) 
  -> smt2 -> z3 (not ok? stop)
  -> ok -> pyrs -> rust code
 ```
 
 This has the same benefit that [mojo is claiming](https://docs.modular.com/mojo/manual/parameters/#:~:text=To%20support%20Modular's%20work%20in,just%20a%20few%20new%20features.).

**Reply by adsharma** [reply]
To summarize the thoughts and links above into a more coherent paragraph: There are 3 sub languages here.

* The language the programmer is using to solve the task
* Compile time metaprogramming (C++ templates, rust macros)
* Language for contracts (verus-lang, f-star)

If these three things require the developer to learn three different sublanguages the [cognitive load](https://minds.md/zakirullin/cognitive) will be so high that the execution of the original task suffers.

The proposal here is that we use one language to do all three (although we might use different, increasingly smaller subsets of python). Mojo uses one language for the first two tasks and thinks it's an improvement.

The disadvantage of course is that we need to keep them in one repo and coupling will be tighter vs alternative designs. [verus-lang](https://github.com/verus-lang/verus) lives in a different repo from the rest of rust.

**Comment by jayvdb** [comment]
Created https://github.com/py2many/py2many/issues/568

**Comment by adsharma** [comment, marked as answer]
> Wouldn't ensuring that opposites of assertions are unsatisfiable be much more meaningful?

Yes. Addressed in #602

**Comment by adsharma** [comment]
Thought a bit about the other point @jayvdb raised in this thread. How do we make it valid python?

```
def equation(x: int, y: int):
    assert x > 2
    assert y < 10
    assert x + 2 * y == 7
```

is valid python and more representative of how this feature is likely to be used anyway.

The smt backend needs to handle pre/post conditions as well as the body of a potentially large python function where the post conditions could be violated and then z3 should be able to come up with a counter example explaining to the user how that condition was violated. So it's not just a language for translating assertions. Needs to handle everyday python code as well.

The rest of the stuff (imports, check_sat() etc) could be injected by the transpiler using rewrites.   

But yeah - smt will be an odd ball backend. Unlike other backends, we can't verify that python code and transpiled code produce the same output.

**Reply by jayvdb** [reply]
Have you looked through https://github.com/jayvdb/smtfe ?

Using `assert` would be an improvement IMO, but I was trying to stay close to the Python syntax you had chosen.

**Reply by adsharma** [reply]
I did. From this [example:](https://github.com/jayvdb/smtfe/blob/main/tests/test_formula.py#L84-L85)

```
        a = Variable("a")
        b = Variable("b")
```

This seems to be using the same ideas as z3py and sympy. My bet is that users aren't going to code like this. There is a higher chance that we can convince them to write 

```
def demorgan(a: bool, b: bool):
...
````

Suppose there is a 50 line python function we're trying to verify, users will need to scan the code and extract:

- all the variables being used in that code to create `Variable()` representations.
- then express their asserts in terms of these symbolic representations, which are different from the runtime representations of those variables
- then transpile them to (a) verification language such as smt2 (b) runtime language such as rust

The key is to unify the verification language and runtime language. Otherwise we run into the problem prevalent in distributed consensus space that I've spent some time in. Someone writes a raft TLA+ model, verifies that it is correct. But the actual production code diverges from the model and bugs get introduced.

Here's another more realistic case to think about:

```
def foo(x: int, y: int):
   if smt.pre:
       assert x > 10 and x < 20
       assert y > 20 and y < 30
   if smt.post:
       assert ret > 30 and ret < 50
   return x + y
```

If someone changes the last line to:

```
   return x + 2 * y
```   

We need to get a z3 failure with a counter example (x=19, y = 29, ret = 77). Now repeat the above example with a more complex function body with function calls, lists and other commonly used python constructs.
